examples/hdf5.py

```diff
@@ -4,99 +4,177 @@
 
 The basic approach is to override PlotCurveItem.viewRangeChanged such that it
-reads only the portion of the HDF5 data that is necessary to display the visible
-portion of the data. This is further downsampled to reduce the number of samples 
-being displayed.
-
-A more clever implementation of this class would employ some kind of caching 
-to avoid re-reading the entire visible waveform at every update.
+displays only the visible portion of downsampled data, plus some padding to reduce the frequency
+of plot updates. Further, a relatively small amount of downsampled data is precomputed and stored
+in the HDF5 file, which is sufficient for navigation of the entire data set without aliasing or lag.
 """
 
-import initExample ## Add path to library (just for examples; you do not need this)
+import initExample  # Add path to library (just for examples; you do not need this)
+
+import os
+import sys
 
-import sys, os
-import numpy as np
 import h5py
+import numpy as np
 import pyqtgraph as pg
 from pyqtgraph.Qt import QtCore, QtGui
 
 pg.mkQApp()
 
-
 plt = pg.plot()
 plt.setWindowTitle('pyqtgraph example: HDF5 big data')
 plt.enableAutoRange(False, False)
 plt.setXRange(0, 500)
 
-class HDF5Plot(pg.PlotCurveItem):
+
+class BigDataPlot(pg.PlotCurveItem):
     def __init__(self, *args, **kwds):
-        self.hdf5 = None
-        self.limit = 10000 # maximum number of samples to be plotted
+        self._plotData = None
         pg.PlotCurveItem.__init__(self, *args, **kwds)
-        
-    def setHDF5(self, data):
-        self.hdf5 = data
-        self.updateHDF5Plot()
-        
+
+    def setPlotData(self, plotData):
+        self._plotData = plotData
+        self.replot()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.replot(lazy=True)
+
+    def replot(self, lazy=False):
+        if self._plotData is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
+
+        x1, x2 = vb.viewRange()[0]
+        data = self._plotData.sample(x1, x2, lazy=lazy)
+
+        if data is not None:
+            self.setData(*data)  # update the plot
+
+
+class PlotData:
+    def __init__(self, x, y, cache=None):
+        if x is None:
+            x = len(y)
+        elif len(x) != len(y):
+            raise ValueError
+
+        self._x = x
```

Comment:
doesn't seem like either of these are getting set anywhere besides here.
i'm guessing it's needed for compatibility inside `PlotCurveItem`?